elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/GsonHttpEntity.java

```diff
@@ -0,0 +1,335 @@
+
+	private static byte[] NEWLINE = "\n".getBytes( StandardCharsets.UTF_8 );
+
+	private static final BasicHeader CONTENT_TYPE = new BasicHeader( HTTP.CONTENT_TYPE, ContentType.APPLICATION_JSON.toString() );
+
+	private final Gson gson;
+	private final List<JsonObject> bodyParts;
+
+	/**
+	 * We don't want to compute the length in advance as it would defeat all optimisations.
+	 * Still it's possible that we happen to find out, for example if a Digest from all
+	 * content needs to be computed.
+	 */
+	private long contentLength = -1l;
+
+	/**
+	 * Since flow control might hint to stop producing data,
+	 * while we can't interrupt the rendering of a single JSON body
+	 * we can avoid starting the rendering of any subsequent JSON body.
+	 * So keep track of the next body which still needs to be rendered;
+	 * to allow the output to be "repeatable" we also need to reset this
+	 * at the end.
+	 */
+	private int nextBodyToEncodeIndex = 0;
+
+	/**
+	 * Adaptor from string output rendered into the actual output sink.
+	 * We keep this as a field level attribute as we might have
+	 * partially rendered JSON stored in its buffers while flow control
+	 * refuses to accept more bytes.
+	 */
+	private final ProgressiveByteBufWriter writer = new ProgressiveByteBufWriter();
+
+	public GsonHttpEntity(Gson gson, List<JsonObject> bodyParts) {
+		Objects.requireNonNull( gson );
+		Objects.requireNonNull( bodyParts );
+		this.gson = gson;
+		this.bodyParts = bodyParts;
+	}
+
+	@Override
+	public boolean isRepeatable() {
+		return true;
+	}
+
+	@Override
+	public boolean isChunked() {
+		return false;
+	}
+
+	@Override
+	public long getContentLength() {
+		return contentLength;
+	}
+
+	@Override
+	public Header getContentType() {
+		return CONTENT_TYPE;
+	}
+
+	@Override
+	public Header getContentEncoding() {
+		//Apparently this is the correct value:
+		return null;
+	}
+
+	@Override
+	public InputStream getContent() throws IOException {
+		//This could be implemented but would be sub-optimal compared to using produceContent().
+		//We therefore prefer throwing the exception so that we can easily spot unintended usage via tests.
+		throw new UnsupportedOperationException( "Not implemented! Expected to produce content only over produceContent()" );
+	}
+
+	@Override
+	public void writeTo(OutputStream outstream) throws IOException {
+		//This could be implemented but would be sub-optimal compared to using produceContent().
+		//We therefore prefer throwing the exception so that we can easily spot unintended usage via tests.
+		throw new UnsupportedOperationException( "Not implemented! Expected to produce content only over produceContent()" );
+	}
+
+	@Override
+	public boolean isStreaming() {
+		return false;
+	}
+
+	@Override
+	public void consumeContent() throws IOException {
+		//not used (and deprecated)
+	}
+
+	@Override
+	public void close() throws IOException {
+		//Nothing to close
```

Comment:
that's not possible: the async http client expects to be able to call us multiple times and resume the data stream from the last time, so I can't just reset for the sake of it.